### DISCUSSION TITLE
why time revalidate need request twice to get the fresh data?

### DISCUSSION THREAD

**Discussion by kanbykanby** [vercel/next.js#69685] [App Router] [answered]
more than the specific revalidate seconds, when a request comes, nextjs just return the stale data immediately and fetch the fresh data, then update the cache.

because of that, user need to refresh twice even the duration is more that the revalidate seconds. This is not good for some scenarios. I want a time based revalidation but how to make user get fresh data with only one refresh?
![image](https://github.com/user-attachments/assets/fee40850-2912-49af-99f6-ebe91ff2eaee)

**Comment by Juneezee** [comment, marked as answer]
This is the popular cache invalidation behavior called stale-while-revalidate, see https://datatracker.ietf.org/doc/html/rfc5861#section-3.

It's a design choice made by the Next.js team. If it doesn't fit your scenario then you'll have to find a third-party solution

**Reply by kanbykanby** [reply]
Very appreciate.
It would be great if Next.js allowed me to configure the default cache behavior. Now it's more like a black box.

**Comment by joulev** [comment]
If you don't want the stale-while-revalidate behaviour, you can use `unstable_cache`

```tsx
import { unstable_cache } from "next/cache";

async function getData() {
  const revalidate = 10;
  // This `key` will change every 10 seconds
  const key = Math.floor(new Date().valueOf() / (1000 * revalidate)).toString();
  return unstable_cache(async () => {
    // await db.select().from(...)
    await new Promise((resolve) => setTimeout(resolve, 2000));
    return new Date().toISOString();
  }, [key])();
}

export default async function Page() {
  const data = await getData();
  return <div>{data}</div>;
}

// Make the page dynamic, so that getData() always get called.
// Only the getData() gets called every request, the expensive data query is cached.
export const revalidate = 0;
```

**Reply by kanbykanby** [reply]
Cool! Thank you for this inspiration! Do you mean I'm on my own with `unstable_cache`?
Another question is: If I use the default cache behavior and then ship the app on Vercel, I think this fetch cache is in the local `memory`. How does Vercel sync these caches on different pods?

**Reply by joulev** [reply]
@shezhangzhang `unstable_cache` is a public, documented API from Next.js, meaning it’s not internal and you can use it. It is just unstable, meaning you have to look out for potential breaking changes in non-major versions.

Without `unstable_cache`, to my knowledge it’s not possible to get out of the stale-while-revalidate behaviour in your case.

Where the cache is stored is platform specific and you need not worry about this. `unstable_cache` and `revalidate` both use the same data cache so internal behaviours on where and how caches are stored should be the same for both.

**Comment by Gavrushuk** [comment]
Seems like I have the same issue with revlidate in Next.js 15, I tried to change stale time from 300s to 60s, but it doesn't help.
I opened a discussion with details https://github.com/vercel/next.js/discussions/86538.
Please check, maybe you will have some suggestions in this case. Thank you!
